sample/samplers.go

```diff
@@ -32,25 +43,34 @@ func (s weighted) Sample(logits []float32) (int32, error) {
 		logits64[i] = float64(v)
 	}
 
-	for _, t := range s.transforms {
-		logits64 = t.Apply(logits64)
-	}
+	probs := softmax(logits64)
 
-	logitsCopy := make([]float64, 0, len(logits))
-	indices := make([]int, 0, len(logits))
-	for i, logit := range logits64 {
-		if !math.IsInf(logit, -1) {
-			logitsCopy = append(logitsCopy, logit)
-			indices = append(indices, i)
+	tokens := make([]tokenInfo, len(logits))
+	for i, v := range logits {
+		tokens[i] = tokenInfo{
+			id:    i,
+			logit: float64(v),
+			prob:  probs[i],
 		}
 	}
 
-	if len(logitsCopy) == 0 {
+	tokensInfo := tokenSliceInfo{tokens: tokens, sorted: false}
+	for _, t := range s.transforms {
+		tokensInfo = t.Apply(tokensInfo)
+	}
+
+	if len(tokensInfo.tokens) == 0 {
 		return -1, errors.New("no valid logits found for weighed sampling")
 	}
 
-	probs := softmax(logitsCopy)
-	w := sampleuv.NewWeighted(probs, s.src)
+	filteredProbs := make([]float64, len(tokensInfo.tokens))
+	indices := make([]int, len(tokensInfo.tokens))
```

Comment:
i have a couple of these allocations and would like to remove them - please send ideas 🙏🏽